### DISCUSSION TITLE
Full growatt emulation

### DISCUSSION THREAD

**Discussion by geoffles** [johanmeijer/grott#66] [Ideas]
I'd like to investigate the feasibilty of doing full emulation between the growatt sever and the inverters.

The reason for this is is because the Growatt server seems to be down for a whole day every week and its annoying - I personally would like to have the full details coming through all day, every day, and not just the buffered values - and also to be able to send commands to the inverter even when they cannot provide that to me.

Given the wealth of knowledge around the protocol here, I'm hoping that we can capitalise on that to make this a reality.

Below is an illustration of what I am thinking.

![Grottemu](https://user-images.githubusercontent.com/870474/129188060-a1f29afa-e3aa-4071-9925-7af006ecc1da.png)

A key difference in the design would be that the two TCP connections would be completely independent and that data would flow between the two using a queue - where as currently (correct me if I'm wrong) the local socket writes straight to the server socket and both are torn down if either breaks.

Thoughts and comments?

**Comment by johanmeijer** [comment]
@geoffles I have not noticed that the the Growatt server has been down for a day. But it can be true from your region. 

Do you really want the data also been sent to the Growatt server? Otherwise a simple (? I still haven't found out the total protocol) server emulating the growatt server might be enough (still on my todo list) . 

Grott is already able to store data in a influxdb and that can be used to serve the get data's from an other clients (like Grafana). 

If you really want to sent also data to the Growatt server. This seems to be a good solution. I am not sure what do you mean with the command forward. If you want commands being forwarded from the Growatt server  (e.g. set ip addresses, inverter firmware upgrades from Growatt) I think that the 2 step approach might be challenge and only will work in near real time. But there might be a risk to corrupt the inverter / shine stick (hopefully it is protect against protocol errors). And to be honest I build in the posibillity to filter these commands in Grott to protect your installation against the "threat" from China but  Grott will block then also firmware upgrades.    

The basic protocol is described here: https://www.vromans.org/johan/software/sw_growatt_wifi_protocol.html
But this a little bit outdated. Now a days the data is scrambled (with a simple scramble routine) and the protocol has been  extended because of new inverters and new features.

**Reply by geoffles** [reply]
Interesting that you don't seem to have the same outage problem.

I get your point about the command forwarding - in short, the main reason is because I don't want a service listening on the internet for remote commands. For example, my home automation stuff piggy backs off of HomeKit which then manages the security of getting messages onto my home network. In the case of the inverters, you could also use the shine app if you don't have some other home automation gateway.

Possible corruption is of course my biggest concern - last thing you want to is change the output settings of the inverter! 😄
